Makefile

```diff
@@ -8,95 +50,240 @@ include $(ROOT)/deps/readies/mk/main
+
+CMAKE_FILES= \
+	src/CMakeLists.txt \
+	src/VecSim/spaces/CMakeLists.txt \
+	cmake/common.cmake \
+	cmake/gtest.cmake \
+	cmake/clang-sanitizers.cmake
+
+ifeq ($(DEBUG),1)
+CMAKE_BUILD_TYPE=DEBUG
+else
+CMAKE_BUILD_TYPE=RelWithDebInfo
+endif
+CMAKE_FLAGS += -DCMAKE_BUILD_TYPE=$(CMAKE_BUILD_TYPE)
+
+ifeq ($(VERBOSE),1)
+CMAKE_FLAGS += -DCMAKE_VERBOSE_MAKEFILE=on
+endif
+
+CMAKE_FLAGS += \
+	-Wno-deprecated \
+	-DCMAKE_WARN_DEPRECATED=OFF \
+	-Wno-dev \
+	--no-warn-unused-cli \
+	-DOSNICK=$(OSNICK) \
+	-DARCH=$(ARCH) \
+	$(CMAKE_SAN) \
+	$(CMAKE_VECSIM)
+
+#----------------------------------------------------------------------------------------------
 
 include $(MK)/defs
-include $(MK)/rules
 
 #----------------------------------------------------------------------------------------------
 
-all: build
+all: bindirs $(TARGET)
 
 .PHONY: all
 
+include $(MK)/rules
+
+#----------------------------------------------------------------------------------------------
+
+.PHONY: __force
+
+$(BINDIR)/Makefile: __force
+	$(SHOW)cd $(BINDIR) && cmake $(CMAKE_FLAGS) $(CMAKE_DIR)
+
+$(TARGET): $(BINDIR)/Makefile
+	@echo Building $(TARGET) ...
+ifneq ($(DRY_RUN),1)
+	$(SHOW)$(MAKE) -C $(BINDIR) $(MAKE_J)
+else
+	@make -C $(BINDIR) $(MAKE_J)
+endif
+
+clean:
+ifeq ($(ALL),1)
+	$(SHOW)rm -rf $(BINROOT) build dist .tox
+else
+	$(SHOW)$(MAKE) -C $(BINDIR) clean
+endif
+
+.PHONY: clean
+
+#----------------------------------------------------------------------------------------------
+
+pybind:
+	$(SHOW)python3 -m poetry build
+
 #----------------------------------------------------------------------------------------------
 
+_CTEST_ARGS=$(CTEST_ARGS)
+_CTEST_ARGS += \
+	--output-on-failure \
+	$(MAKE_J)
+
 ifeq ($(VERBOSE),1)
-CMAKE_FLAGS+=-DCMAKE_VERBOSE_MAKEFILE=on
+_CTEST_ARGS += -V
+endif
+
+ifeq ($(VALGRIND),1)
+_CTEST_ARGS += \
+	-T memcheck \
+	--overwrite MemoryCheckCommandOptions="--leak-check=full"
 endif
 
-build:
-	mkdir -p build
-	cd build; cmake $(CMAKE_FLAGS) ../src
-	-touch build/Makefile
-	make -C build
+unit_test:
+	$(SHOW)cd $(BINDIR)/unit_tests && GTEST_COLOR=1 ctest $(_CTEST_ARGS)
 
-clean:
-ifneq ($(ALL),1)
-	make -C build clean
+#----------------------------------------------------------------------------------------------
+
+flow_test:
+ifneq ($(VIRTUAL_ENV),)
+	$(SHOW)cd tests/flow && python3 -m pytest $(TEST)
 else
-	rm -rf build
+	$(SHOW)$(MAKE) pybind
+	$(SHOW)python3 -m tox -e flowenv
 endif
 
-.PHONY: build clean
+#----------------------------------------------------------------------------------------------
+
+ifneq ($(TEST),)
+RLTEST_ARGS += --test $(TEST) -s
+endif
+ifeq ($(VERBOSE),1)
+RLTEST_ARGS += -v
+endif
+
+mod_test:
+	$(SHOW)cd $(ROOT)/tests/module/flow && \
+		python3 -m RLTest \
+		--module $(BINDIR)/module_tests//memory_test.so \
+		--clear-logs \
+		$(RLTEST_ARGS)
 
 #----------------------------------------------------------------------------------------------
 
-test:
-	./tests/test.sh
+benchmark:
```

Comment:
call build first, please, as we will use it during benchmark development 
also, there might be more than one benchmark, so please modify it with the ability to call multiple benchmarks

Reply:
I usually separate build and test because when you run tests repeatedly it's annoying to see the build run in vain. The most frequent changes will be to the python files anyway.